### ISSUE TITLE
OIP-1 Pause Gaurdian

### ISSUE THREAD

**Issue by DanielVF** [OriginProtocol/origin-dollar#984]
There is often a window of 3-7 minutes between an attack contract deployment and the execution of the attack. These attacks are often first noticed by people external to the organization. To pause currently requires:

1. Someone to notice the exploit
2. Find out how to contact us
3. Actually contact someone
4. Convince someone
5. Initial origin responder has to contact another signer
6. We have to create the pause transaction.
7. Second signer has to sign

Instead we could create a contract that has one time use pause actions. After 12 hours OUSD automatically unpauses unless the pause is confirmed by a multisig. These pause authorizations can then be distributed to external security researchers that are monitoring the blockchain. Then if they see an attack come through they can pause immediately in one step. This could cut the reaction time down to 2-3 minutes.

Even if this was only used internally, this would still be quiet the speed difference.

**Comment by franckc** [comment]
I like the idea since it drastically reduces the incident response time.
I assume we would whitelist wallet addresses of the security researchers that would be allowed use this functionality?

**Comment by DanielVF** [comment]
Yes, it would be to specific trusted individuals.

**Comment by sparrowDom** [comment]
Great idea! Does it make sense for this to become an EIP? So that contracts to be paused all implement a common interface. Should be easier for security researchers to interact with those. Especially in such time critical situations. 

I can totally see a need for a service (with website similar to snapshot.org) where other protocols aside from OUSD opt in and give permissions to specific security researchers to pause their contracts. And previous activity of security researchers could be seen, so that trust can be based on historical success. So it is easier for platforms to pick who to trust.

Naturally there should also be monetary incentives when an attack is prevented.

**Comment by DanielVF** [comment]
We have two pause methods, though I wonder if we should combine them into one.

**Comment by staa99** [comment]
Hi @franckc,

The idea is interesting and I'd like to implement the contract.

Considering the requirement that the caller of the pause functions, `pauseRebase` and `pauseCapital`, must be the `governor` or `strategist`, then it might be necessary to extend the `governor` or `strategist` with the desired functionality or upgrade the contract to allow a new address (the `pauser`) to pause/unpause the vault.

What do you think?

**Comment by franckc** [comment]
> Considering the requirement that the caller of the pause functions, `pauseRebase` and `pauseCapital`, must be the `governor` or `strategist`, then it might be necessary to extend the `governor` or `strategist` with the desired functionality or upgrade the contract to allow a new address (the `pauser`) to pause/unpause the vault.

I'd think we would want to leave the current `governor` and `strategist` roles as is. But we would add a new role that perhaps we call `pauser` that is made up of a whitelist of addresses. The `governor`, `strategist` and `pauser` would all have permissions to call the pause methods.

**Comment by staa99** [comment]
`pauser` can be one address (the `pauser` contract), which manages the whitelist of security researchers that can use the function. The pausing rules can then be implemented in the `pauser` contract. The access requirements for the `pause/unpause` functions can then be updated to allow the single `pauser` contract.

**Comment by DanielVF** [comment]
Yes, I agree with you @staa99. One pause address the vault, with the rest managed by the pause contract.

**Comment by staa99** [comment]
Hi @franckc @DanielVF,

I have a few questions on the `pauser` contract.

- Who can unpause after a confirmed pause? (maybe the governor?)
- Can the `vaultAdmin` address be changed? (maybe yes, but only by the governor?)
- Can the `expiryDuration` be changed? (maybe yes, but only by the governor?)

**Comment by staa99** [comment]
Thanks a lot for clarifying @DanielVF

**Comment by staa99** [comment]
Hi @DanielVF,

One last question before I submit a PR. Thanks a lot for your help on this issue.

`VaultAdmin` defines an `unpauseRebase` method which can only be called by the governor.

https://github.com/OriginProtocol/origin-dollar/blob/bf1ba296d1efdbc5491f1f029a0ab6e4c84cbe74/contracts/contracts/vault/VaultAdmin.sol#L285

The tests for the method include the following:

https://github.com/OriginProtocol/origin-dollar/blob/bf1ba296d1efdbc5491f1f029a0ab6e4c84cbe74/contracts/test/vault/rebase.js#L46-52

The test is described as `Should allow strategist to unpause rebasing`, however, it tests for the exact opposite (the strategist cannot unpause rebasing and will instead revert). From the description of the strategist powers you gave above, it seems to me that the test and modifier are incorrect and should be updated. If the test is correct and the current behavior is desired, the pauser should not have access to unpause rebasing either, since the strategist can unpause with the pauser.

Please clarify.

**Comment by DanielVF** [comment]
Thank you for noticing. The strategist should indeed be able to to unpause rebasing. (We had changed our minds a few times in the past.)

**Comment by staa99** [comment]
Hi @DanielVF,

Please review my PR when you're available. Thanks a lot for your time and help.
